lib/vector-common/src/request_metadata.rs

```diff
@@ -1,16 +1,180 @@
+use std::collections::HashMap;
 use std::ops::Add;
 
+use crate::internal_event::{CountByteSize, OptionalTag};
 use crate::json_size::JsonSize;
 
+/// (Source, Service)
+pub type EventCountTags = (OptionalTag, OptionalTag);
+
+/// Must be implemented by events to get the tags that will be attached to
+/// the `component_sent_event_*` emitted metrics.
+pub trait GetEventCountTags {
+    fn get_tags(&self) -> EventCountTags;
+}
+
+/// Keeps track of the estimated json size of a given batch of events by
+/// source and service.
+#[derive(Clone, Debug)]
+pub enum RequestCountByteSize {
```

Comment:
Can we call this `GroupedCountByteSize`? "Request" doesn't really indicate what it's doing over `CountByteSize`.